v2/protocol/http/protocol_retry.go

```diff
@@ -79,51 +78,34 @@ func (p *Protocol) doWithRetry(ctx context.Context, params *cecontext.RetryParam
 
 		// Fast track common case.
 		if protocol.IsACK(result) {
-			return msg, NewRetriesResult(result, retry, then, results)
+			return msg, NewRetriesResult(result, retry, start, results)
 		}
 
-		// Try again?
-		//
-		// Make sure the error was something we should retry.
-
-		{
-			var uErr *url.Error
-			if errors.As(result, &uErr) {
-				goto DoBackoff
+		var httpResult *Result
+		if errors.As(result, &httpResult) {
+			sc := httpResult.StatusCode
+			if !p.isRetriableFunc(sc) {
+				cecontext.LoggerFrom(ctx).Debugw("status code not retryable, will not try again",
+					zap.Error(httpResult),
+					zap.Int("statusCode", sc))
+				return msg, NewRetriesResult(result, retry, start, results)
 			}
 		}
 
-		{
-			var httpResult *Result
-			if errors.As(result, &httpResult) {
-				sc := httpResult.StatusCode
-				if p.isRetriableFunc(sc) {
-					// retry!
-					goto DoBackoff
-				} else {
-					// Permanent error
-					cecontext.LoggerFrom(ctx).Debugw("status code not retryable, will not try again",
-						zap.Error(httpResult),
-						zap.Int("statusCode", sc))
-					return msg, NewRetriesResult(result, retry, then, results)
-				}
-			}
-		}
-
-	DoBackoff:
-		resetBody(req, body)
-
-		// Wait for the correct amount of backoff time.
-
 		// total tries = retry + 1
-		if err := params.Backoff(ctx, retry+1); err != nil {
+		if err = params.Backoff(ctx, retry+1); err != nil {
 			// do not try again.
 			cecontext.LoggerFrom(ctx).Debugw("backoff error, will not try again", zap.Error(err))
-			return msg, NewRetriesResult(result, retry, then, results)
+			return msg, NewRetriesResult(result, retry, start, results)
 		}
 
 		retry++
+		resetBody(req, body)
```

Comment:
Overall this looks ok to me but I have a question.... more of a golang question... below in `resetBody` we'll reset the req.Body appropriately, but then only set req.GetBody once, if ever, based on whether it's already set or not. Now, it seems to me that the Reader created on line 124 (new numbering) is never reset - so that would then imply that GetBody() is never actually called during a retry. Do I have the right? If so, do we need lines 121-126 at all?

Reply:
Tbh, I don't fully recall why I had to add these lines as part of https://github.com/cloudevents/sdk-go/issues/773 but IIRC it was because of a `nil` pointer panic during tests with `httptest` using `GetBody`. Looks like our code base is not using `GetBody` at all, just setting it in `v2/protocol/http/write_request.go`, with a `default` case not setting `GetBody`. 

IMHO it's correct to only set it once (ever) since it's caller e.g., `httptest`, then must handle the case of re-reading a previously read body. But I could be totally wrong :)